### PR TITLE
ENG-969: [PXC/PXB/ProxySQL/IaC] Implement AWS ECR for our internal build images

### PR DESCRIPTION
**Pull request by Sudokamikaze**
Link to prepare job: https://ps3.cd.percona.com/job/prepare-pxc-build-docker-ecr/

Before merge tasks:
* Remove unneeded pointer to my local repo for prepare job
* Test usage build (from PS, uses the same way of IAM):  https://ps80.cd.percona.com/job/percona-server-8.0-pipeline-ecrtest/5/

### CODE REVIEW

File: IaC/JenkinsImagesECR.yml [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+          Statement:
+          - Effect: Allow
+            Resource:
+            - '*'    
+            Action:
+            - ecr-public:GetAuthorizationToken
+            - sts:GetServiceBearerToken
+      Tags: 
+        - Key: iit-billing-tag
+          Value: jenkins-ecr-readwriteuser
```
Comment (delgod) [inline IaC/JenkinsImagesECR.yml:72]: please ```suggestion
          Value: jenkins-ps-ecr
```

File: pxc/jenkins/prepare-pxc-build-docker.yml [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+                label 'docker-32gb'
+            }
+            options {
+                skipStagesAfterUnstable()
+                buildDiscarder(logRotator(artifactNumToKeepStr: '10'))
+            }
+            stages {
+                stage('Prepare') {
+                    steps {
+                        git poll: true, branch: 'ENG-969', url: 'https://github.com/Sudokamikaze/jenkins-pipelines'
```
Comment (delgod) [inline pxc/jenkins/prepare-pxc-build-docker.yml:45]: please use our repo

File: pxc/jenkins/prepare-pxc-build-docker.yml [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+                sg docker -c "
+                    ./pxc/docker/prepare-docker ${SOURCE_IMAGE}
+                "
+            """
+            withCredentials([[$class: 'AmazonWebServicesCredentialsBinding', accessKeyVariable: 'AWS_ACCESS_KEY_ID', credentialsId: 'ECRRWUser', secretKeyVariable: 'AWS_SECRET_ACCESS_KEY']]) {
+                sh """
+                    SOURCE_IMAGE=${SOURCE_IMAGE}
+                    ECR_TEMP_PASS="\$(aws ecr-public get-login-password --region us-east-1)"
+                    sg docker -c "
+                        docker login -u AWS -p \$ECR_TEMP_PASS public.ecr.aws/e7j3v3n0
```
Comment (delgod) [inline pxc/jenkins/prepare-pxc-build-docker.yml:28]: more secure command (password not shown in the logs)

```
aws ecr-public get-login-password --region us-east-1 \
    | docker login --username AWS --password-stdin public.ecr.aws/e7j3v3n0
```

**Review by delgod — CHANGES_REQUESTED** [review]
`./pxc/jenkins/pxc57-pipeline.groovy,./pxc/jenkins/pxc56-pipeline.groovy,./pxc/jenkins/pxc80-pipeline.groovy,./pxb/docker/pxb24-single-platform-run.groovy,./pxb/docker/pxb80-single-platform-run.groovy`  must do `docker login`.

reason: when docker image is downloaded using an AWS Account - traffic is free.
please see https://aws.amazon.com/ecr/pricing/

File: IaC/JenkinsImagesECR.yml [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    Properties:
+      RepositoryName: "ps-build"
+      RepositoryCatalogData:
+        UsageText: "This image used exclusively for Percona Jenkins pipelines"
+        AboutText: "This image contains all build utilities to create proper builds and execute tests"
+        OperatingSystems:
+          - "Linux"
+        Architectures:
+          - "x86"
+          - "x86-64"
```
[the remark this answers is not in the document]
Comment (Sudokamikaze) [reply IaC/JenkinsImagesECR.yml:19]: This resource type can't be tagged: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ecr-publicrepository.html

You're viewing wrong doc, ecr and ecr-public are different resources

File: pxc/jenkins/prepare-pxc-build-docker.yml (hunk 2 above)
Comment (Sudokamikaze) [reply pxc/jenkins/prepare-pxc-build-docker.yml:45]: Will be implemented before merge

File: IaC/JenkinsImagesECR.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (Sudokamikaze) [reply IaC/JenkinsImagesECR.yml:19]: Also, it is not possible to add tags via Web GUI and via CLI, there's no such property

File: IaC/JenkinsImagesECR.yml [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+          - Effect: Allow
+            Resource:
+            - '*'    
+            Action:
+            - ecr-public:GetAuthorizationToken
+            - sts:GetServiceBearerToken
+      Tags: 
+        - Key: iit-billing-tag
+          Value: jenkins-ecr
+      UserName: "ECRROUser"
```
Comment (delgod) [inline IaC/JenkinsImagesECR.yml:109]: please it is critical to be able to identify resources from the name
```suggestion
          Value: jenkins-ps-ecr
      UserName: "jenkins-ps-ecr"
```

**Review by delgod — CHANGES_REQUESTED** [review]

File: pxc/jenkins/pxc56-pipeline.groovy [hunk 6]
```diff
@@ -200,7 +202,8 @@ pipeline {
                             until aws s3 cp --no-progress s3://pxc-build-cache/${BUILD_TAG}/garbd ./pxc/sources/pxc56/garbd; do
                                 sleep 5
                             done
-							
+                            
+							aws ecr-public get-login-password --region us-east-1 | docker login -u AWS --password-stdin public.ecr.aws/e7j3v3n0
```
Comment (delgod) [inline pxc/jenkins/pxc56-pipeline.groovy:206]: wrong tabulation?

File: pxc/jenkins/pxc57-pipeline.groovy [hunk 7]
```diff
@@ -165,7 +166,8 @@ pipeline {
 
                     echo 'Build PXC57'
                     withCredentials([[$class: 'AmazonWebServicesCredentialsBinding', accessKeyVariable: 'AWS_ACCESS_KEY_ID', credentialsId: 'c42456e5-c28d-4962-b32c-b75d161bff27', secretKeyVariable: 'AWS_SECRET_ACCESS_KEY']]) {
-                        sh '''
```
Comment (delgod) [inline pxc/jenkins/pxc57-pipeline.groovy:168]: why this line changed?

File: pxc/jenkins/qa-pxc57-pipeline.groovy [hunk 8]
```diff
@@ -146,7 +147,8 @@ pipeline {
                     
                         echo 'Build PXC56'
                         withCredentials([[$class: 'AmazonWebServicesCredentialsBinding', accessKeyVariable: 'AWS_ACCESS_KEY_ID', credentialsId: 'c42456e5-c28d-4962-b32c-b75d161bff27', secretKeyVariable: 'AWS_SECRET_ACCESS_KEY']]) {
-                            sh '''
```
Comment (delgod) [inline pxc/jenkins/qa-pxc57-pipeline.groovy:149]: why it is changed?

File: pxc/jenkins/qa-pxc57-pipeline.groovy [hunk 9]
```diff
@@ -193,7 +195,9 @@ pipeline {
                     
                         echo 'Build PXC56'
                         withCredentials([[$class: 'AmazonWebServicesCredentialsBinding', accessKeyVariable: 'AWS_ACCESS_KEY_ID', credentialsId: 'c42456e5-c28d-4962-b32c-b75d161bff27', secretKeyVariable: 'AWS_SECRET_ACCESS_KEY']]) {
-                            sh '''							
+                            sh '''			
+
```
Comment (delgod) [inline pxc/jenkins/qa-pxc57-pipeline.groovy:199]: wrong tabulation?

**Review by delgod — CHANGES_REQUESTED** [review]

File: pxc/jenkins/pxc56-pipeline.groovy (hunk 6 above)
Comment (Sudokamikaze) [reply pxc/jenkins/pxc56-pipeline.groovy]: Corrected

File: pxc/jenkins/pxc57-pipeline.groovy (hunk 7 above)
Comment (Sudokamikaze) [reply pxc/jenkins/pxc57-pipeline.groovy:168]: Corrected

File: pxc/jenkins/qa-pxc57-pipeline.groovy (hunk 8 above)
Comment (Sudokamikaze) [reply pxc/jenkins/qa-pxc57-pipeline.groovy:149]: Corrected

File: pxc/jenkins/qa-pxc57-pipeline.groovy (hunk 9 above)
Comment (Sudokamikaze) [reply pxc/jenkins/qa-pxc57-pipeline.groovy]: Corrected

**Review by delgod — APPROVED** [review]

**Review by EvgeniyPatlan — APPROVED** [review]

**Pull request merged**